collections/index.js

```diff
@@ -20,46 +20,130 @@ import employees from './data.json';
-
+  // created an empty object to insert data into
+  var employeeNumber = {};
+  // using a for loop to index through each of the employees on record
+  for (let employeeNumberIndex = 0; employeeNumberIndex < employees.length; employeeNumberIndex++) {
+    // in each of the iterations, if the employee property 'id' matches the
+    // provided arguement 'id', the statement is true...
+    if (employees[employeeNumberIndex].id === id) {
+      // ... and the following code is executed:
+      // this assigns the employees details matching their id to the
+      // variable 'employeeNumber'
+      employeeNumber = employees[employeeNumberIndex];
+    }
+  }
+  // this returns the employee details as the result of the function.
+  return employeeNumber;
 }
 
 /*
   I expect this function to return me an array of employees depending on which
   department is passed in
 */
 export const getEmployeesByDepartment = (dept) => {
-
+  // sets an empty array ready to accept any values we add later:
+  var departmentEmployees = [];
+  // using a for loop to index through each of the employees on record
+  for (let employeeNumberIndex = 0; employeeNumberIndex < employees.length; employeeNumberIndex++) {
+    // in each of the iterations, if the employee 'department' matches the
+    // provided arguement 'dept', the statement is true...
+    if (employees[employeeNumberIndex].department === dept) {
+      // this adds (pushes) the employees details to the departmentEmployees
+      // array we created earlier.
+      departmentEmployees.push(employees[employeeNumberIndex]);
+    }
+  }
+  // returns all employees that match the department or an empty array if
+  // there were no matches.
+  return departmentEmployees;
+  // side note: i wrote this completely before testing and it worked first time!
 }
 
 /*
   I expect this function to look through all the employees and return an array
   of employees who are missing any data
 */
-export const getEmployeesWithMissingData = () => {
 
+export const getEmployeesWithMissingData1 = () => {
+  // sets an empty array ready to accept any values we add later:
+  var missingInfoEmployees = [];
+  // using a for loop to index through each of the employees on record
+  for (let employeeNumberIndex = 0; employeeNumberIndex < employees.length; employeeNumberIndex++) {
+    // checks if any of the values are empty. I prefer this method over others
+    // I thought of because if more than one value is empty per employee, this
+    // function only adds the employee to the array once.
+    if (employees[employeeNumberIndex].id === "" || employees[employeeNumberIndex].firstName === "" || employees[employeeNumberIndex].lastName === "" || employees[employeeNumberIndex].department === "" || employees[employeeNumberIndex].contactNumber === "" || employees[employeeNumberIndex].taxCode === "" || employees[employeeNumberIndex].salary === "") {
+      // this adds (pushes) the employees details to the missingInfoEmployees
```

Comment:
Remember the D.R.Y principle - you're accessing `employees[employeeNumberIndex]` a lot in this check. You can cast it to a variable at the start of the loop and just access it like that

```js
for (...) {
  let emp = employees[index];
  if (emp.id === '' ...) {
    arr.push()
  }
}
```

Reply:
Also, you can check for an empty string but what if the data is missing completely (i.e. `undefined`?) - this check would actually fail because 

```js
console.log(undefined === '') // false
```

A more generalised way of checking the absence of something is to check if the value is falsey

```js
if (!employee.id || !employee.firstName) {

}
```

Reply:
I thought of that, but then I remember you mentioning that one of the data types in the future might be true or false and that would break in future.

Reply:
I thought of that, but then I remember you mentioning that one of the data types in the future might be true or false and that would break in future.